src/communication_managers/iothub_communication_manager/src/iothub_communication_manager.c

```diff
@@ -0,0 +1,714 @@
+ *
+ */
+static void ADUC_Refresh_IotHub_Connection_SAS_Token()
+{
+    if (g_aduc_client_handle_address == NULL)
+    {
+        Log_Error("Invalidate operation. Must call IoTHub_CommunicationManager_Init() to initialize the manager.");
+        return;
+    }
+
+    if (g_aduc_client_handle_address != NULL && *g_aduc_client_handle_address != NULL)
+    {
+        ADUC_DeviceClient_Destroy(*g_aduc_client_handle_address);
+        *g_aduc_client_handle_address = NULL;
+        if (g_iothub_client_handle_changed_callback != NULL)
+        {
+            g_iothub_client_handle_changed_callback(*g_aduc_client_handle_address);
+        }
+    }
+
+    ADUC_ConnectionInfo info = {};
+    if (!GetAgentConfigInfo(&info))
+    {
+        goto done;
+    }
+
+    if (!ADUC_DeviceClient_Create(g_aduc_client_handle_address, &info, true /* iotHubTracingEnabled */))
+    {
+        Log_Error("ADUC_DeviceClient_Create failed");
+        goto done;
+    }
+
+    if (g_iothub_client_handle_changed_callback != NULL)
+    {
+        g_iothub_client_handle_changed_callback(*g_aduc_client_handle_address);
+    }
+
+    Log_Info("Successfully re-authenticated the IoT Hub connection.");
+
+done:
+
+    ADUC_ConnectionInfo_DeAlloc(&info);
+}
+
+/**
+ * @brief Performs an authentication to the IoTHub as needed, with exponential back-off retry logics.
+ *
+ */
+static void Connection_Maintenance()
+{
+    if (IoTHub_CommunicationManager_IsAuthenticated())
+    {
+        return;
+    }
+
+    // Try to (re)connect to the IoT Hub if:
+    //   1. The connection is broken (or unauthenticated)
+    //   2. It has been long enough since the last authentication attemps
+    time_t now_time = GetTimeSinceEpochInSeconds();
+
+    if (now_time < g_next_authentication_attempt_time)
+    {
+        return;
+    }
+
+    // If we haven't tried to connect, no need to compute the next retry time.
+    // Otherwise, compute next retry time we've attempted to authenticate after the previous time.
+    if (g_last_authentication_attempt_time != 0 &&
+        g_last_authentication_attempt_time >= g_next_authentication_attempt_time)
+    {
+        // Calculate the next retry time, then continue.
+        time_t nextRetryTime = ADUC_Retry_Delay_Calculator(
+                TIME_SPAN_FIFTEEN_SECONDS_IN_SECONDS,
+                g_authentication_retries /* current retires count */,
+                ADUC_RETRY_DEFAULT_INITIAL_DELAY_MS /* initialDelayUnitMilliSecs */,
+                TIME_SPAN_ONE_HOUR_IN_SECONDS,
+                ADUC_RETRY_DEFAULT_MAX_JITTER_PERCENT);
+
+        g_next_authentication_attempt_time = (nextRetryTime);
+        Log_Info("The connection is currently broken. Will try to authenticate in %d seconds.", nextRetryTime - now_time);
+        return;
+    }
+
+    // Try to authenticate.
+    g_last_authentication_attempt_time = now_time;
+    g_authentication_retries++;
+    ADUC_Refresh_IotHub_Connection_SAS_Token();
```

Comment:
This isn't productive for connections that are made manually. Did I miss a check for the type of connection we're trying to make? 

In the end it shouldn't matter but if we're seeing a "sas token expired" on a connection that is using shared access keys (eg manual connection strings) we should err out or make that log. That's an unexpected case that should be handled by the C-SDK and not the agent at large.  

That could almost be a bug or a log line instead of a larger feature change.

Reply:
As discussed, for the 'string' connection type, when the authentication broken, we should try to re-connect, in case that the root cause is fixed by the solution operator.